token_test.go

```diff
@@ -1056,6 +1058,43 @@ var _ = Describe("Tokens", func() {
 			_, _, err = connection.Tokens()
 			Expect(err).To(HaveOccurred())
 		})
+
+		It("Honours context timeout", func() {
+			// Generate tokens:
+			refreshToken := DefaultToken("Refresh", 10*time.Hour)
+			accessToken := DefaultToken("Bearer", 5*time.Minute)
+
+			// Configure the server with a handler that introduces an
+			// artificial delay:
+			oidServer.AppendHandlers(
+				ghttp.CombineHandlers(
+					http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+						time.Sleep(10 * time.Millisecond)
+					}),
+					VerifyRefreshGrant(refreshToken),
+					RespondWithTokens(accessToken, refreshToken),
+				),
+			)
+
+			// Create the connection:
+			connection, err := NewConnectionBuilder().
+				Logger(logger).
+				Metrics(metrics).
+				TokenURL(oidServer.URL()).
+				URL(apiServer.URL()).
+				Tokens(refreshToken).
+				Build()
+			Expect(err).ToNot(HaveOccurred())
+			defer connection.Close()
+
+			// Request the token with a timeout smaller than the artificial
+			// delay introduced by the server:
+			ctx, _ := context.WithTimeout(context.Background(), 5*time.Millisecond)
+			_, _, err = connection.TokensContext(ctx)
+
+			// The request should fail:
+			Expect(err).To(HaveOccurred())
```

Comment:
Can we check the error to make sure it's due to conext deadline reached?

Reply:
Yes, can be checked. It requires changing calls to `fmt.Errorf` to use `%w` instead of `%v` for errors. I am doing that in #289 . Will update this once that is merged.